File: src/main/java/org/lanternpowered/porygen/settings/json/DoublePredicateParser.java
/*
 * This file is part of Porygen, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.porygen.settings.json;

import com.google.common.base.Joiner;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import org.apache.commons.lang3.StringUtils;
import org.lanternpowered.porygen.util.RegexPatterns;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.DoublePredicate;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * Parses a {@link DoublePredicate}.
 * <p>Examples:
 * </p>
 * <p>
 * This example will match any number between 10 and 20 (inclusive)
 * >= 10 and <= 20
 * </p>
 * <p>
 * This example will match any number between 10 and 20 (inclusive), except 15
 * (>= 10 and <= 20) and not = 15
 * </p>
 * <p>
 * This example will match 7 or 9
 * = 9 or = 7
 * </p>
 */
public class DoublePredicateParser implements JsonDeserializer<DoublePredicate> {

    /**
     * All the match operators that are supported.
     */
    private final Map<String, MatchOperatorProvider> matchOperatorProviders = new HashMap<>();

    /**
     * All the join operators that are supported.
     */
    private final Map<String, JoinOperatorProvider> joinOperatorProviders = new HashMap<>();

    /**
     * The pattern to match a single {@link DoublePredicate} operation.
     */
    private final Pattern matchOperatorPattern;

    /**
     * The pattern to match a single join operation.
     */
    private final Pattern joinOperatorPattern;

    /**
     * The pattern to match content between parentheses.
     */
    private final Pattern parenthesesPattern = Pattern.compile("\\([^(]*\\)");

    {
        // Create the operators
        this.matchOperatorProviders.put(">", value -> (v -> v > value));
        this.matchOperatorProviders.put("<", value -> (v -> v < value));
        this.matchOperatorProviders.put("<=", value -> (v -> v <= value));
        this.matchOperatorProviders.put(">=", value -> (v -> v >= value));
        this.matchOperatorProviders.put("=", value -> (v -> v == value));
        this.matchOperatorProviders.put("!=", value -> (v -> v != value));
        this.matchOperatorProviders.put("<>", value -> (v -> v != value));

        this.joinOperatorProviders.put("and", DoublePredicate::and);
        this.joinOperatorProviders.put("or", DoublePredicate::or);
        this.joinOperatorProviders.put("and not", (a, b) -> a.and(b.negate()));
        this.joinOperatorProviders.put("or not", (a, b) -> a.or(b.negate()));

        // Create the pattern based on the registered operators
        String operators = Joiner.on('|').join(this.matchOperatorProviders.keySet());
        this.matchOperatorPattern = Pattern.compile("(?:(" + operators + ")\\s*(" + RegexPatterns.DOUBLE_PATTERN + ")|@([0-9]+))");

        operators = Joiner.on('|').join(this.joinOperatorProviders.keySet().stream()
                .sorted((a, b) -> b.length() - a.length())
                .map(s -> s.replace(" ", "\\s+"))
                .collect(Collectors.toList()));
        this.joinOperatorPattern = Pattern.compile("\\s+(" + operators + ")\\s+");
    }

    @Override
    public DoublePredicate deserialize(
            JsonElement element, Type type, JsonDeserializationContext ctx) throws JsonParseException {
        String conditions = element.getAsString();
        if (conditions.equalsIgnoreCase("true")) {
            return v -> true;
        } else if (conditions.equalsIgnoreCase("false")) {
            return v -> false;
        }

        // Parsed sub conditions
        final List<DoublePredicate> parsedConditions = new ArrayList<>();

        Matcher matcher;
        while ((matcher = this.parenthesesPattern.matcher(conditions)).find()) {
            // Get the condition without the parentheses
            final String condition = conditions.substring(matcher.start() + 1, matcher.end() - 1);
            final int index = parsedConditions.size();
            parsedConditions.add(parse(condition, parsedConditions));
            // Replace the section with the index
            conditions = matcher.replaceFirst("@" + index);
        }

        return parse(conditions, parsedConditions);
    }

    private DoublePredicate parse(String condition, List<DoublePredicate> parsedConditions) {
        DoublePredicate parsedCondition = null;
        while (true) {
            JoinOperatorProvider joinOperatorProvider = null;
            if (parsedCondition != null) {
                final Matcher joinMatcher = this.joinOperatorPattern.matcher(condition);
                if (!joinMatcher.find()) {
                    throw new IllegalArgumentException("Invalid join operator: " + condition);
                }
                condition = condition.substring(joinMatcher.end());
                if (condition.trim().isEmpty()) {
                    throw new IllegalArgumentException("Expected another condition");
                }
                final String operator = StringUtils.normalizeSpace(joinMatcher.group(1));
                joinOperatorProvider = this.joinOperatorProviders.get(operator);
            }
            final Matcher matchMatcher = this.matchOperatorPattern.matcher(condition);
            if (!matchMatcher.find()) {
                throw new IllegalArgumentException("Invalid condition: " + condition);
            }
            condition = condition.substring(matchMatcher.end());
            // Get the operator
            final String operator = matchMatcher.group(1);
            final DoublePredicate parsedCondition1;
            if (operator == null) {
                // References a other parsed condition
                parsedCondition1 = parsedConditions.get(Integer.parseInt(matchMatcher.group(3)));
            } else {
                final double value = Double.parseDouble(matchMatcher.group(2));
                final MatchOperatorProvider matchOperatorProvider = this.matchOperatorProviders.get(operator);
                parsedCondition1 = matchOperatorProvider.get(value);
            }
            if (joinOperatorProvider == null) {
                parsedCondition = parsedCondition1;
            } else {
                parsedCondition = joinOperatorProvider.join(parsedCondition, parsedCondition1);
            }
            if (condition.trim().isEmpty()) {
                break;
            }
        }
        return parsedCondition;
    }

    @FunctionalInterface
    private interface MatchOperatorProvider {

        DoublePredicate get(double value);
    }

    @FunctionalInterface
    private interface JoinOperatorProvider {

        DoublePredicate join(DoublePredicate a, DoublePredicate b);
    }
}

File: src/main/kotlin/org/lanternpowered/porygen/parser/types/DoublePredicateParser.kt
/*
 * This file is part of Porygen, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.porygen.parser.types

import com.google.common.base.Joiner
import com.google.common.reflect.TypeToken
import org.apache.commons.lang3.StringUtils
import org.lanternpowered.porygen.parser.ParseException
import org.lanternpowered.porygen.parser.PoryElement
import org.lanternpowered.porygen.parser.PoryParser
import org.lanternpowered.porygen.parser.PoryParserContext
import org.lanternpowered.porygen.util.RegexPatterns
import java.util.*
import java.util.function.DoublePredicate
import java.util.regex.Matcher
import java.util.regex.Pattern
import kotlin.streams.toList

typealias MatchOperatorProvider = (Double) -> DoublePredicate
typealias JoinOperatorProvider = (DoublePredicate, DoublePredicate) -> DoublePredicate

/**
 * Parses a [DoublePredicate].
 *
 * Examples:
 *
 * This example will match any number between 10 and 20 (inclusive)
 * >= 10 and <= 20
 *
 * This example will match any number between 10 and 20 (inclusive), except 15
 * (>= 10 and <= 20) and not = 15
 *
 * This example will match 7 or 9
 * = 9 or = 7
 */
class DoublePredicateParser : PoryParser<DoublePredicate> {

    /**
     * All the match operators that are supported.
     */
    private val matchOperatorProviders = HashMap<String, MatchOperatorProvider>()

    /**
     * All the join operators that are supported.
     */
    private val joinOperatorProviders = HashMap<String, JoinOperatorProvider>()

    /**
     * The pattern to match a single [DoublePredicate] operation.
     */
    private val matchOperatorPattern: Pattern

    /**
     * The pattern to match a single join operation.
     */
    private val joinOperatorPattern: Pattern

    /**
     * The pattern to match content between parentheses.
     */
    private val parenthesesPattern = Pattern.compile("\\([^(]*\\)")

    init {
        // Create the operators
        this.matchOperatorProviders[">"] = { value -> DoublePredicate { v -> v > value } }
        this.matchOperatorProviders["<"] = { value -> DoublePredicate { v -> v < value } }
        this.matchOperatorProviders["<="] = { value -> DoublePredicate { v -> v <= value } }
        this.matchOperatorProviders[">="] = { value -> DoublePredicate { v -> v >= value } }
        this.matchOperatorProviders["="] = { value -> DoublePredicate { v -> v == value } }
        this.matchOperatorProviders["!="] = { value -> DoublePredicate { v -> v != value } }
        this.matchOperatorProviders["<>"] = { value -> DoublePredicate { v -> v != value } }

        this.joinOperatorProviders["and"] = { obj, other -> obj.and(other) }
        this.joinOperatorProviders["or"] = { obj, other -> obj.or(other) }
        this.joinOperatorProviders["and not"] = { a, b -> a.and(b.negate()) }
        this.joinOperatorProviders["or not"] = { a, b -> a.or(b.negate()) }

        // Create the pattern based on the registered operators
        var operators = Joiner.on('|').join(this.matchOperatorProviders.keys)
        this.matchOperatorPattern = Pattern.compile("(?:($operators)\\s*(${RegexPatterns.DOUBLE_PATTERN})|@([0-9]+))")

        operators = Joiner.on('|').join(this.joinOperatorProviders.keys.stream()
                .sorted { a, b -> b.length - a.length }
                .map { s -> s.replace(" ", "\\s+") }
                .toList())
        this.joinOperatorPattern = Pattern.compile("\\s+($operators)\\s+")
    }

    override fun parse(element: PoryElement, type: TypeToken<DoublePredicate>, ctx: PoryParserContext): DoublePredicate {
        var conditions = element.asString()
        if (conditions.equals("true", ignoreCase = true)) {
            return DoublePredicate { true }
        } else if (conditions.equals("false", ignoreCase = true)) {
            return DoublePredicate { false }
        }

        // Parsed sub conditions
        val parsedConditions = ArrayList<DoublePredicate>()

        var matcher: Matcher
        while (true) {
            matcher = this.parenthesesPattern.matcher(conditions)
            if (!matcher.find()) break
            // Get the condition without the parentheses
            val condition = conditions.substring(matcher.start() + 1, matcher.end() - 1)
            val index = parsedConditions.size
            parsedConditions.add(parse(condition, parsedConditions))
            // Replace the section with the index
            conditions = matcher.replaceFirst("@$index")
        }

        return parse(conditions, parsedConditions)
    }

    private fun parse(condition: String, parsedConditions: List<DoublePredicate>): DoublePredicate {
        var conditionValue = condition
        var parsedCondition: DoublePredicate? = null
        while (true) {
            var joinOperatorProvider: JoinOperatorProvider? = null
            if (parsedCondition != null) {
                val joinMatcher = this.joinOperatorPattern.matcher(conditionValue)
                if (!joinMatcher.find()) {
                    throw IllegalArgumentException("Invalid join operator: $conditionValue")
                }
                conditionValue = conditionValue.substring(joinMatcher.end())
                if (conditionValue.trim { it <= ' ' }.isEmpty()) {
                    throw IllegalArgumentException("Expected another condition")
                }
                val operator = StringUtils.normalizeSpace(joinMatcher.group(1))
                joinOperatorProvider = this.joinOperatorProviders[operator]
            }
            val matchMatcher = this.matchOperatorPattern.matcher(conditionValue)
            if (!matchMatcher.find()) {
                throw IllegalArgumentException("Invalid condition: $conditionValue")
            }
            conditionValue = conditionValue.substring(matchMatcher.end())
            // Get the operator
            val operator = matchMatcher.group(1)
            val nextParsedCondition = if (operator == null) {
                // References a other parsed condition
                parsedConditions[Integer.parseInt(matchMatcher.group(3))]
            } else {
                val value = java.lang.Double.parseDouble(matchMatcher.group(2))
                val matchOperatorProvider = this.matchOperatorProviders[operator] ?: throw ParseException("Unknown operator: $operator")
                matchOperatorProvider(value)
            }
            parsedCondition = joinOperatorProvider?.invoke(parsedCondition!!, nextParsedCondition) ?: nextParsedCondition
            if (conditionValue.trim { it <= ' ' }.isEmpty()) {
                break
            }
        }
        return parsedCondition!!
    }
}
